File: NeetCode/Greedy/Jump_Game/jump__game.py
from typing import Optional, List


class Solution:
    def canJump(self, nums: List[int]) -> bool:
        endGoal = len(nums) - 1

        for index in range(len(nums) - 2, -1, -1):
            if index + nums[index] >= endGoal:
                endGoal = index

        return endGoal == 0


if __name__ == "__main__":
    sol = Solution()
    nums = [1, 2, 0, 1, 0]
    print(sol.canJump(nums))
    print("Running Solution...")

File: NeetCode/Greedy/Jump_Game/Jump_Game.cpp
// Jump_Game.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    bool canJump(vector<int> &nums)
    {
        int endgoal = nums.size() - 1;
        for (int index = endgoal - 1; index > -1; index--)
        {
            if (index + nums[index] >= endgoal)
            {
                endgoal = index;
            }
        }
        return endgoal == 0;
    }
};
